File: app/src/main/java/com/example/weather/search/SearchFragment.java
package com.example.weather.search;

import androidx.lifecycle.ViewModelProvider;

import android.content.pm.PackageManager;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.example.weather.databinding.FragmentSearchBinding;
import com.example.weather.helpers.LocationHelper;
import com.example.weather.helpers.SharedPreferenceHelper;

public class SearchFragment extends Fragment {
    private SearchViewModel viewModel;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        FragmentSearchBinding binding = FragmentSearchBinding.inflate(inflater);
        viewModel = new ViewModelProvider(this).get(SearchViewModel.class);

        binding.setViewModel(viewModel);
        binding.setLifecycleOwner(getViewLifecycleOwner());

        /*
        * If there was a city in shared preferences then load the weather for it
        * Otherwise check if we have location enabled and load the weather for that
        * Otherwise request the permission
        * */
        String city = SharedPreferenceHelper.getCity(requireContext());
        if (!city.isEmpty()) {
            viewModel.updateLocationWithCity(city);
        } else if (LocationHelper.checkPermission(requireContext())) {
            LocationHelper.getLocation(requireContext(), viewModel::updateLocation);
        } else {
            requestPermissions(LocationHelper.permissions, LocationHelper.LOCATION_PERMISSION_REQUEST);
        }

        return binding.getRoot();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == LocationHelper.LOCATION_PERMISSION_REQUEST) {
            for (int i = 0; i < permissions.length; i++) {
                if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                    //Tell the user to go to settings when permission is denied
                    if (!shouldShowRequestPermissionRationale(permissions[i])) {
                        Toast.makeText(requireContext(), "Go to settings and enable the permission", Toast.LENGTH_SHORT).show();
                    }
                } else {
                    //if the permission request was successful then get the weather based on it
                    LocationHelper.getLocation(requireContext(), viewModel::updateLocation);
                }
            }
        }
    }
}

File: app/src/main/java/com/example/weather/search/SearchFragment.kt
package com.example.weather.search

import android.content.pm.PackageManager
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.viewModels
import com.example.weather.databinding.FragmentSearchBinding
import com.example.weather.helpers.LocationHelper
import com.example.weather.helpers.LocationHelper.PERMISSION_REQUEST
import com.example.weather.helpers.SharedPreferenceHelper

class SearchFragment : Fragment() {
    private val viewModel: SearchViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val binding = FragmentSearchBinding.inflate(inflater)
        binding.viewModel = viewModel
        binding.lifecycleOwner = viewLifecycleOwner

        context?.let { appContext ->
            val city = SharedPreferenceHelper.getCity(appContext)
            if (!city.isNullOrEmpty()) {
                viewModel.updateLocationWithCity(city)
            } else if (LocationHelper.checkPermission(appContext)) {
                LocationHelper.getLocation(appContext) {
                    viewModel.updateLocation(it)
                }
            } else{
                requestPermissions(LocationHelper.permissions, PERMISSION_REQUEST)
            }
        }

        return binding.root
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSION_REQUEST) {
            for (i in permissions.indices) {
                if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                    if (shouldShowRequestPermissionRationale(permissions[i])) {
                        Toast.makeText(context, "Permission denied", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(context, "Go to settings and enable the permission", Toast.LENGTH_SHORT).show()
                    }
                } else {
                    context?.let { appContext ->
                        LocationHelper.getLocation(appContext) {
                            viewModel.updateLocation(it)
                        }
                    }

                }
            }
        }
    }
}